rust/datafusion/src/optimizer/utils.rs

```diff
@@ -98,117 +98,117 @@ pub fn exprlist_to_fields(expr: &Vec<Expr>, input_schema: &Schema) -> Result<Vec
 
 /// Given two datatypes, determine the supertype that both types can safely be cast to
 pub fn get_supertype(l: &DataType, r: &DataType) -> Result<DataType> {
-    match _get_supertype(l, r) {
-        Some(dt) => Ok(dt),
-        None => match _get_supertype(r, l) {
-            Some(dt) => Ok(dt),
-            None => Err(ExecutionError::InternalError(format!(
-                "Failed to determine supertype of {:?} and {:?}",
-                l, r
-            ))),
-        },
+    use arrow::datatypes::DataType::*;
+    let d = if l == r {
```

Comment:
In the old implementation the super type of `UInt16` and `Int8` is `None` but now it is `Int16`. Why is that? Also do you know how C++ handles this case?

Reply:
It is not statically compatible in C++ (safely). https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernels/cast.cc#L156-L208

It can be safe to cast, but it needs a runtime check.